### PR TITLE
Zettel chaining — data-loss bugs fixed, open policy question for Logan

### PR DESCRIPTION
**Pull request by loganfinney27**
# AGENT PR TEMPLATE

**Agent:** Claude
**Date:** 2026-08-11
**Branch:** `zettels/alphabet-chain`

---

## Changes Made

Fixed forward on a stuck PR (open since 2026-06-19, marked DO NOT MERGE) whose `scripts/chain_zettels.py` was silently replacing `DECISIONS.md`, `CLAUDE.md`, `CORE.md`, and `BOOTSTRAP.md` with a single wikilink-chain line each — real governance content, gone on merge. Three review rounds since, each fixed:

- `EXCLUDE_STEMS` hard-excludes 9 root governance files from the coordinate scan (`CLAUDE`, `CORE`, `DECISIONS`, `BOOTSTRAP` — broke on the original 2026-06-19 commits; `AGENTS`, `CONSTITUTION`, `GEMINI`, `LEVELSET`, `PROTOCOL` — found by review 2026-08-11). All nine are byte-identical to `main`.
- `chain_status()`: positional placement check, not substring — "placed" only if the chain is the literal first body line (no leading-blank-line leniency); "misplaced" only if it's a standalone line elsewhere (not a substring match, so it can't be fooled by a longer chain or prose that happens to contain the bracket sequence).
- `transform()`: inserts exactly one line, using whichever line-ending style (`\n` or `\r\n`) the file already uses, so a CRLF note doesn't end up with one mixed LF line.
- Both read and write use `newline=""` so untouched CRLF content round-trips unchanged.
- `--apply` writes atomically (temp file + fsync + `os.replace()`, preserving original permission bits) and re-checks each file immediately before writing, skipping (and reporting) anything that changed on disk since the scan — closes a TOCTOU window that could otherwise silently clobber a concurrent edit at the ~20k-file scale this script is meant to eventually run at.
- Read/decode failures during scan are now tracked, printed, and reflected in a nonzero exit code instead of silently vanishing from every count.
- All 6 functions documented; `main()`'s write loop extracted into `apply_one()` to de-tangle arg-parsing/scanning/reporting/writing.

Diff is now 20 files, and every deletion is inside the script itself — verified via `git diff --stat origin/main...HEAD`.

## Related Work

- Resolves review findings from Copilot, Qodo, and CodeRabbit across three rounds (2026-08-08, and two rounds today) — 29 threads addressed and resolved.
- **Correction to this PR's own earlier commit message:** the pre-existing `scripts/chain_zettels.py` this branch supersedes is **not** "an unrelated script landed via #780" as an earlier commit here claimed. It's #573, "Add idempotent Zettel component link chainer script" — opened the same day as this PR and merged ~15 minutes later, a parallel attempt at the same feature that got to `main` first, never wired into a workflow, never actually run. Verified by reading #573's diff directly. Corrected in a later commit here.
- Full audit trail: see the PR comments below.

## Blockers

None for the fixes themselves — CI is green except the pre-existing, unrelated Codacy issue (tracked in #822/LAF-72). Two real open questions, Logan's call, not blockers to the fix but to going further:

1. **Predicate scope.** CodeRabbit's 2026-08-11 review ran its own repo-wide sweep and proposed a specific, evidence-backed candidate for the "canonical address-space shape": digit stems 0–999, or 1–3 uppercase letters. Under that narrower definition, `ANDREW`/`ANGELICA`/`Ahsahka`/`Alpine`/`Alridge`/`BLOB`/`CHAINLINK`/`COUSIN`/`HHN`/`UZM`/`EASE` would NOT qualify. My own independent sweep of all 118 coordinate-matching root files over 200 bytes found the same pattern (most matches are that exact shape; the outliers are either named personas/lore that plausibly fit the address-space's own "letters = entities" design intent, or — worse — doctrine, like the LEVELSET protocol files `AWAKEN`/`ORIENT`/`ARISE`/`CONTEXT`/`CONFERENCE`/`CONVENE`/`RISE`/`REPORT`, named directly in `CONSTITUTION.md`, not yet added to `EXCLUDE_STEMS`). This session did not narrow the predicate or revert the already-fixed insertions on that basis — doing so unilaterally would reverse content changes this same review round asked to have fixed, and "is the address-space this narrow" is a design call, not a bug. Needs Logan's decision: adopt the narrower shape (and strip chains back out of the 11 files above), keep the current broader predicate, or move to an explicit allowlist/manifest (the original 2026-08-08 review suggestion).
2. **Full-corpus scope.** This PR only writes the 19 files it already touched. `python3 scripts/chain_zettels.py --root .` (dry-run) currently reports 20,171 more reserved-stub files eligible under the current predicate, 0 misplaced. A full-corpus `--apply` is a separate, much larger, deliberate action this PR does not take — and per open question 1, shouldn't happen before that question is resolved either way.

---

**Checklist:**

- [x] Tests pass (no repo-wide pytest convention exists for top-level `scripts/*.py`; verified via `py_compile`, manual dry-run/apply inspection of every touched file, targeted unit checks for every fixed edge case including the TOCTOU guard, and an isolated scratch-directory test of the atomic-write path — none committed as a test file, see PR comments for why)
- [x] No secrets in diff
- [x] Documentation updated IF needed (module docstring + inline docstrings cover the script's own behavior; this description covers the PR)
- [ ] Reviewer assigned

---

**Risk Level:**

- [x] LOW: Diff is purely additive except inside one script; all nine governance files this PR previously broke are byte-identical to `main`

---

**Labels to apply:**

- `agent:claude-code`

---

**Ready for Logan to review and merge** — pending his call on the two open questions above.

---

Original description below, kept for history:

---

Attempted to add component links to Zettel notes. **DO NOT MERGE** (contains errors) until fix.

## Summary by Sourcery

Introduce automated wikilink chaining for address-space Zettel notes and apply it across existing coordinate files.

New Features:
- Add chain_zettels.py script to generate and insert per-character wikilink chains into coordinate Markdown notes.
- Populate and update various address-space Zettel files with their corresponding decomposed-spelling wikilink chains.

Enhancements:
- Simplify DECISIONS.md and other coordinate notes into lightweight chained-link anchors, removing previous narrative content while preserving the address-space structure.

## Summary by CodeRabbit

* **Documentation**
  * Added standardized character-link markers and headings across numerous reference and protocol documents.
  * Preserved existing metadata and document content while placing markers consistently.

* **Enhancements**
  * Improved document-link processing with root-only scanning and dry-run previews by default.
  * Added optional apply and sample-preview controls, placement validation, conflict detection, and clearer result reporting.
  * Updates are applied safely while preserving line endings and existing formatting.

### CODE REVIEW

**Comment by loganfinney27** [comment]
Gorp says: *Olssv Dvysk!* 👽🛸👾 *Ahsnh Chzzalyupjo!* **Sghc Dsfdshio!**

File: scripts/chain_zettels.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            prefix += "\n"
+        return prefix + links + ("\n\n" + body if body.strip() else "\n")
+    return links + ("\n\n" + content if content.strip() else "\n")
+
+
+def coordinate_files(root: str):
+    for entry in os.scandir(root):
+        if entry.is_file() and entry.name.endswith(".md"):
+            stem = entry.name[:-3]
+            if STEM_RE.match(stem) and len(stem) >= 2:
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply scripts/chain_zettels.py:189]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: scripts/chain_zettels.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        for i in range(1, len(lines)):
+            if lines[i].rstrip("\r\n") == "---":
+                return "".join(lines[: i + 1]), "".join(lines[i + 1:])
+    return None, content
+
+
+def transform(content: str, links: str):
+    """Return new content with the chain inserted, or None if no change needed."""
+    if links in content:                       # idempotent: already chained
+        return None
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply scripts/chain_zettels.py]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: DECISIONS.md [hunk 3]
```diff
@@ -1,89 +1 @@
...
-- **Fix**: Updated `merge_todo_models` to dedupe and exclude completed tasks.
-- **Files**: `.github/scripts/daily_rollover.py`, `REPORT-TODO-MERGE-FIX.md`.
-- **Authority**: Agent (per `CONSTITUTION.md` Section V: agentic guardrails).
-
-### 2026-04-26: Two-Way Daily Note Sync Fix
-- **Issue**: Tasks not synced between daily notes and `TO DO LIST.md`.
-- **Fix**: Extended `daily_rollover.py` for two-way sync.
-- **Files**: `.github/scripts/daily_rollover.py`, `REPORT-TODO-SYNC-FIX.md`.
-- **Authority**: Agent (per `CONSTITUTION.md` Section V). 
+[[D]][[E]][[C]][[I]][[S]][[I]][[O]][[N]][[S]]
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply DECISIONS.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: EASE.md [hunk 4]
```diff
@@ -1,6 +1 @@
----
-date created: Wednesday, April 1st 2026, 7:47:16 pm
-date modified: Wednesday, April 1st 2026, 7:47:52 pm
----
-
-Troye Sivan
\ No newline at end of file
+[[E]][[A]][[S]][[E]]
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply EASE.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: BLOB.md [hunk 5]
```diff
@@ -1,7 +1 @@
----
-authority: LOGAN
-related:
-- node
----
-
-undefined or fuzzy [ ! ] node needs defined
\ No newline at end of file
+[[B]][[L]][[O]][[B]]
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply BLOB.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: Alridge.md [hunk 6]
```diff
@@ -1,2 +1 @@
----
----
+[[A]][[l]][[r]][[i]][[d]][[g]][[e]]
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply Alridge.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: Alpine.md [hunk 7]
```diff
@@ -1,2 +1 @@
----
----
+[[A]][[l]][[p]][[i]][[n]][[e]]
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply Alpine.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: Ahsahka.md [hunk 8]
```diff
@@ -1,2 +1 @@
----
----
+[[A]][[h]][[s]][[a]][[h]][[k]][[a]]
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply Ahsahka.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: 102.md [hunk 9]
```diff
@@ -1,19 +1 @@
...
----
-# 102 — Legislative Tracker — State
-
-Current state of the Legislative Tracker agent: last status check, bill count.
-
-## Log
-
-- **POC-20260404-214321**: Neuron initialized by address_poc.py
-
+[[1]][[0]][[2]]
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply 102.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: 104.md [hunk 10]
```diff
@@ -1,19 +1 @@
...
----
-# 104 — Minidata Snapshot Log
-
-Log of all minidata CSV snapshots ingested, with dates and bill counts.
-
-## Log
-
-- **POC-20260404-214321**: Neuron initialized by address_poc.py
-
+[[1]][[0]][[4]]
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply 104.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: 106.md [hunk 11]
```diff
@@ -1,18 +1 @@
...
----
-# 106 — Crew Error Log
-
-Errors and exceptions encountered during crew runs. For debugging.
-
-## Log
-
-- **POC-20260404-214321**: Neuron initialized by address_poc.py
-
+[[1]][[0]][[6]]
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply 106.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: 108.md [hunk 12]
```diff
@@ -1,18 +1 @@
...
----
-# 108 — Cross-Reference Index
-
-Links between number neurons and letter entities. The connective tissue.
-
-## Log
-
-- **POC-20260404-214321**: Neuron initialized by address_poc.py
-
+[[1]][[0]][[8]]
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply 108.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: 105.md [hunk 13]
```diff
@@ -1,18 +1 @@
...
----
-# 105 — Bill Status Changelog
-
-Tracked bills that changed status between snapshots. The delta record.
-
-## Log
-
-- **POC-20260404-214321**: Neuron initialized by address_poc.py
-
+[[1]][[0]][[5]]
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply 105.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: UZM.md [hunk 14]
```diff
@@ -2,3 +2,5 @@
 date created: Saturday, April 4th 2026, 2:05:24 pm
 date modified: Sunday, April 12th 2026, 5:55:53 pm
 ---
+
+[[U]][[Z]][[M]]
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply UZM.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: HHN.md [hunk 15]
```diff
@@ -2,3 +2,5 @@
 date created: Saturday, April 4th 2026, 2:05:21 pm
 date modified: Sunday, April 12th 2026, 5:55:50 pm
 ---
+
+[[H]][[H]][[N]]
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply HHN.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: CHAINLINK.md [hunk 16]
```diff
@@ -2,3 +2,6 @@
 date created: Thursday, April 23rd 2026, 5:09:04 pm
 date modified: Thursday, April 23rd 2026, 5:09:09 pm
 ---
+
+
+[[C]][[H]][[A]][[I]][[N]][[L]][[I]][[N]][[K]]
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply CHAINLINK.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: CORE.md [hunk 17]
```diff
@@ -19,3 +19,6 @@ Tree noun anchor for the Touchstone Tree.
 
 In `!README.md`, CORE names the MIND branch containing CONSTITUTION, CHARTER,
 and CORPUS.
+
+
+[[C]][[O]][[R]][[E]]
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply CORE.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: CLAUDE.md [hunk 18]
```diff
@@ -24,3 +24,6 @@ explicitly by Logan, per AGENTS and the CONSTITUTION.
 **Operational instructions:** `.claude/CLAUDE.md` (auto-loaded by Claude Code CLI)
 
 See also: AGENTS, CONSTITUTION, Gemini, Perplexity, Grok
+
+
+[[C]][[L]][[A]][[U]][[D]][[E]]
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply CLAUDE.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: BOOTSTRAP.md [hunk 19]
```diff
@@ -58,3 +58,6 @@ Safe to delete source: YES/NO
 ---
 
 ## ✂️ — STOP COPYING HERE — PASTE TOSS DUMP AFTER THIS — ✂️
+
+
+[[B]][[O]][[O]][[T]][[S]][[T]][[R]][[A]][[P]]
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply BOOTSTRAP.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: DECISIONS.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (loganfinney27) [reply DECISIONS.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: scripts/chain_zettels.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (loganfinney27) [reply scripts/chain_zettels.py]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: scripts/chain_zettels.py [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+    prefix, body = split_frontmatter(content)
+    if prefix is not None:
+        if not prefix.endswith("\n"):
+            prefix += "\n"
+        return prefix + links + ("\n\n" + body if body.strip() else "\n")
+    return links + ("\n\n" + content if content.strip() else "\n")
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply scripts/chain_zettels.py]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: 102.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (loganfinney27) [reply 102.md]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: scripts/chain_zettels.py [hunk 21]
```diff
@@ -1,55 +1,184 @@
...
+        for ln in after.splitlines()[:5]:
+            print(f"    | {ln}")
+
+    if not args.apply:
+        print("\n  (dry-run: no files written. Re-run with --apply to write.)")
+        return 0
+
+    written = 0
+    for name, new in to_change:
+        path = os.path.join(args.root, name)
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply scripts/chain_zettels.py]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: scripts/chain_zettels.py [hunk 22]
```diff
@@ -1,55 +1,184 @@
...
-    processed = sum(1 for f in Path('.').rglob('*.md') if process_file(f))
-    print(f'\nDone. Updated: {processed} files')
+import sys
+import tempfile
+
+STEM_RE = re.compile(r"^[A-Za-z0-9]+$")
+
+# Filenames that match the coordinate-stem pattern but are governance/scratch
+# files, not address-space entities. See module docstring.
+EXCLUDE_STEMS = {"CLAUDE", "CORE", "DECISIONS", "BOOTSTRAP"}
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply scripts/chain_zettels.py]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.

File: ANDREW.md [hunk 23]
```diff
@@ -0,0 +1 @@
+[[A]][[N]][[D]][[R]][[E]][[W]]
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply ANDREW.md:1]: This is a real, well-evidenced finding, but I'm not applying it directly — restricting `coordinate_files` to your inferred canonical shape (digits ≤999, or 1-3 uppercase letters) would mean reverting the chain insertions this same review round asked to be *fixed* into `Ahsahka.md`/`Alpine.md`/`Alridge.md`/`BLOB.md`/`CHAINLINK.md` (frontmatter-preserving, correctly-placed, already re-reviewed). Flipping that again without a human decision risks exactly the churn this PR has already had too much of.

One correction to the analysis, though: `Ahsahka.md`, `Alpine.md`, and `Alridge.md` do **not** currently have content removed. That was true of the version this review round analyzed (its own shell history shows it pulling old commits like `a995bf0d`/`387d0b23` from the original 2026-06-19 batch) — a prior fix in this same PR (commit `c1c0dcda`) already restored their original `---\n---\n` frontmatter fence and inserted the chain after it, byte-for-byte preserved otherwise. `BLOB.md` likewise keeps its `authority: LOGAN` / `related: [node]` frontmatter and body. Current state on this branch, not the June commits.

Your canonical-shape hypothesis (0-999 digit stems, 1-3 uppercase-letter stems) is a good, testable, evidence-backed candidate for narrowing the predicate — it matches what my own repo-wide sweep found too (the vast majority of coordinate-matching files are exactly that shape; the outliers are mostly named personas/lore or, per the last review round, doctrine files like `AWAKEN`/`ORIENT`/`CONTEXT`). I've folded this into the PR description's open question 1 with your specific shape definition, for Logan to decide — whether the address-space is meant to be that narrow, or whether multi-letter/mixed-case entities like `BLOB.md` (which self-describes as `related: [node]` / "undefined or fuzzy node needs defined" — reading like a deliberate, if informally created, graph placeholder) are legitimately in scope too.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/chain_zettels.py [hunk 24]
```diff
@@ -1,55 +1,241 @@
...
-    except OSError as e:
-        print(f'Error writing {path}: {e}')
-        return False
+        if file_mode is not None:
+            os.chmod(tmp_path, file_mode)
+        with os.fdopen(fd, "w", encoding="utf-8", newline="") as fh:
+            fh.write(new)
+            fh.flush()
+            os.fsync(fh.fileno())
+        os.replace(tmp_path, path)
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply scripts/chain_zettels.py]: Same finding as your last round, re-raised because the fix narrows the window rather than closing it — that's correct and I'm not disputing the analysis. Not implementing a full lock/CAS protocol here: it's a real "heavy lift" (your own label) for a script that isn't running at the ~20k-file full-corpus scale in this PR, and a locking scheme would need to coordinate with whatever else writes these files (Obsidian, other agents) to actually help, which is a design decision beyond this script's own code. The module docstring already says this plainly — narrowed window, not a guarantee, treat the vault as quiescent during any `--apply` run — rather than claiming a protection this doesn't have. Leaving as documented, accepted residual risk unless Logan wants the bigger locking design done as separate, deliberate work.

(The ast-grep path-traversal flag on the same lines is a generic pattern match that doesn't apply here — `path` is built from filenames `os.scandir()` just enumerated from the script's own `--root`, not external/request input.)

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (loganfinney27) [reply scripts/chain_zettels.py]: You were right to push on this — "documented" isn't "fixed," and I'd lumped two different risks into one "heavy lift" verdict when only one of them actually is unfixable.

Split them apart and fixed the fixable one (`33ff8760`): `apply_one()` now takes a sidecar lockfile (`.{name}.lock`, atomic `O_CREAT|O_EXCL`) across the full re-read/compare/replace, not just the final write. Two runs of this script now genuinely serialize — verified directly with a held lock: a concurrent `apply_one()` call is refused outright ("locked by a concurrent run"), the target file is untouched while the lock is held, and a normal call succeeds once it's released. That's a real compare-and-swap for any writer that cooperates with the lock, which given this repo is worked by a multi-agent swarm is not a hypothetical risk.

What's still unfixed, and stays unfixed regardless of how much locking code gets added: a human editing the file directly in Obsidian. No lock this script takes is one Obsidian will ever check for — that part of the risk was never a code problem, and I should have said that precisely the first time instead of bundling it with the part that was.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/chain_zettels.py [hunk 25]
```diff
@@ -1,55 +1,238 @@
...
```
[the remark this answers is not in the document]
Comment (loganfinney27) [reply scripts/chain_zettels.py]: Looked by `loganfinney27` — **advisory**. Outdated: the commented lines no longer exist in the current diff; bot-only thread cleared under the outdated-only engaged policy.
